### DISCUSSION TITLE
esp32-s2 MSC fails with partial Read10 and sync reset doesn't seem to be honored

### DISCUSSION THREAD

**Discussion by jewalt** [hathach/tinyusb#1384] [Q&A]
I'm working on a esp32-s2 design where I'm using tiny USB to enable a MSC connection.  If I have USB debugging set to 2 or 3, I never see this error.  When I have USB debug set to 1, on occasion I will see the USB connection start and the MSC intererface enumerate.  When there is a request for a READ10 of LUN0, I see a partial frame where 11 or so pkgs are sent of the expected 64. This is hard to debug as when I turn on debugging the problem goes away, but using wireshark, I see a frame of 731 bytes (sometimes more but the data portion is always a multiple of 64) when it should be 4123 (4096 bytes of data). 

Is it possible that the USB fifo is signaling a complete frame to the far end?  The far end (Windows 10) immediatly responds with a sync Reset and clear stall, but that seems to have no effect.  I'm at a loss of how to debug this problem any further.  Is there someone out there that has seen something like this?

**Comment by jewalt** [comment]
So I have a bit more information. Using a protocol analyzer I can see that with Windows 10, I hit Turnaround/Timeout errors and only a partial packet is sent (less than 64 bytes).  With linux, I don't see this.  Should there be some type of underrun error that could be checked in the FIFO?  After the turnaround timeout, there is no recovery until a bus reset 10s of seconds later.

**Comment by hathach** [comment]
Do you have a reliable way to reproduce the issue ? Which example that you use, if it is custom, please test out the cdc_msc to rule out application bug.

**Comment by jewalt** [comment]
It seems this is an issue with Windows 10.  What I see is Linux is a query for a device qualifier but Windows never does. There is a comment in dcd_es32s2.c that states:
`      // Host sends this request to ask why our device with USB BCD from 2.0
      // but is running at Full/Low Speed. If not highspeed capable stall this request,
      // otherwise return the descriptor that could work in highspeed mode`

What I think is happening is Windows assumes High speed where Linux actually queries.  Since we cannot meet HS timing with ESP32s2, we eventually fall behind on large transfers (hundreds of 64 byte buffers).  So I added a small delay between 64 byte buffers which results in a few NAKs being sent between every data frame.  This seems to allow the ESP32 to get back on time and Windows does not timeout the transfer.  

This seems to be a Windows bug and not a ESP32/tiny USB - MSC bug.

**Reply by hathach** [reply]
windows does ask for device qualifier, it probably caches the result and skip that after 1st plug to save time on enumeration. Regardless of device qualifier present or not. it does not affect the actual speed, bus speed FS/HS is determined in the hard way with actual electrical signal called chirp. So I guess it is probably something else.

If you consistently got timing turn around error, may be you should tweak `gusbcfg` register, there is a field for turn around timing there

**Reply by jewalt** [reply]
Strange, I am not seeing Windows ask for the device qualifier, even after a windows reboot. I will look into the gusbcfg register, but I'm not sure it will help. It clearly appears to be that the ESP32 is not able to keep up when sending back to back data packets. Maybe it is something in my configuration that is causing it, but it works great on Linux (I never see a turnaround error) but it almost always fails on Windows until I added the extra delay to force a NAK between each data packet.
